### ISSUE TITLE
Assign or set by reference to a function of LHS of ':=' in j

### ISSUE THREAD

**Issue by qmarcou** [Rdatatable/data.table#5081]
Hi!
I'm trying to update by reference a function of the LHS of ':=' inside the `j ` argument. I failed to find a corresponding issue or SO thread. #4031 and #795 seem somewhat related but very much tied to the use of column attributes, and I feel what I'm trying to achieve is a more general problem.  
Here is a minimal example of what I'm trying to do:

```r
library(data.table)
library(lubridate)

# Let's build a DT containing a single column with random dates
A=data.table(mydate=as.Date.numeric(runif(100,min = 0,max=10000),origin = "1970-01-01"))

# Let's say for some reason I do not trust the date of the month in my dataset
# I'd like to set them to the first day of the month by reference using lubridate
A[,mday(mydate):=1] # Error in mday(mydate) : object 'mydate' not found
```
Now I realize it may not be clear to the data.table what I'm trying to achieve and I guess `:=` expect a particular syntax to interpret the LHS. `mday(x)<-y` acts as a setter function here, and though a syntax  such as `setmday(x,y)` or `x.setmday(y)` (in other languages with different OOP interfaces than S3 and S4 classes) would be clearer, I'm not sure how I could use them inside DT either. 
The error I get suggests mydate is not evaluated in the `env` of my data.table because it is surrounded by a function call. I guess there is something about R's lazy evaluation here that I do not understand.

I have tried the following alternatives:
```r
A[,mday(get('mydate')):=1] # Error in get("mydate") : object 'mydate' not found

# Using ':='()
A[,':='('mday(mydate)':=1)] # obviously creates a new column named 'mday(mydate)'
A[,':='(mday(get('mydate')):=1)] # Error in `[.data.table`(A, , `:=`(`:=`(mday(get("mydate")), 1))) : In `:=`(col1=val1, col2=val2, ...) form, all arguments must be named.

# Using setattr
A[,setattr(mydate,mday,1)] # Error in name == "names" : comparison (1) is possible only for atomic and list types

# Using a column returned by reference
mday(A[,mydate])<-1 # Error in `[<-.data.table`(`*tmp*`, , mydate, value = c(2312, 212, 8886,  : object 'mydate' not found
```
Of course I could build a copy of the column, perform the operation and assign in back...
```r
tmp <- A[,mydate]
mday(tmp)<-1
A[,mydate:=tmp]
``` 
... however I don't feel that's the spirit of data.table, quite inefficient (and impractical if  I need to perform the operation in a subset using `i`), and risky in case the order in DT has been changed in the meantime.

I'm not very experienced in R and I don't know if having both getter and setter functions in a single function returning an 'attribute' (in the sense of any internal object) by reference is common or not apart for  R attributes.
But in a nutshell my issue is: how can I use a setter function or an assignment to modify an unexposed/private attribute of an object in the LHS of `:=` by reference?
 
Thank so much to the developpers of this great package, it is clearly one of R's strongest assets! 
best regards

`#` `Output of sessionInfo()`
```r
R version 3.6.3 (2020-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/atlas/libblas.so.3.10.3
LAPACK: /usr/lib/x86_64-linux-gnu/atlas/liblapack.so.3.10.3

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=fr_FR.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=fr_FR.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=fr_FR.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=fr_FR.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] lubridate_1.7.9.2 data.table_1.13.6

loaded via a namespace (and not attached):
[1] compiler_3.6.3 generics_0.1.0 tools_3.6.3    Rcpp_1.0.6    
```

**Comment by avimallu** [comment]
`data.table` expects the LHS to be either character(s) or any of the formats prescribed in the documentation for [`:=`](https://www.rdocumentation.org/packages/data.table/versions/1.14.0/topics/%3A%3D). Lubridate's `mday` function does not return the object it works on, and therefore cannot be directly used to assign to the column. Instead, use a function:

```r
mdayfun <- function(dt, val) {
    mday(dt) <- val
    dt
}
A[,mydate:=mdayfun(mydate, val=1)]
```

[You might also want to read this](https://github.com/Rdatatable/data.table/issues/4415), and be careful about using `lubridate` objects in `data.table`, as they don't always play well with each other.

**Comment by Henrik-P** [comment]
You may also use the functional form of the replacement:

    A[ , mydate := `mday<-`(mydate, 1)]

**Comment by qmarcou** [comment]
> `data.table` expects the LHS to be either character(s) or any of the formats prescribed in the [documentation for `:=`](https://www.rdocumentation.org/packages/data.table/versions/1.14.0/topics/%3A%3D). Lubridate's `mday` function does not return the object it works on, and therefore cannot be directly used to assign to the column. Instead, use a function:

Thanks @avimallu for your reply! This is indeed what I had started doing afterwards, and I guess @Henrik-P solution is exactly the same however using dedicated lubridate syntax which is neat! I just had the feeling that using `:=` with the original column name as LHS was not fully efficient as it implies assignment of a modified copy of the column instead of in place modification.

As a side note 
```r
A[ ,  `mday<-`(mydate, 1)]
```
doesn't work as it seems in that case `j` returns a copy of the column and not a reference to it (which makes complete sense since `i` may select random indices and thus the final column cannot be accessed through a single pointer without creating a dedicated array). I guess this is what @avimallu is about. 

> [You might also want to read this](https://github.com/Rdatatable/data.table/issues/4415), and be careful about using `lubridate` objects in `data.table`, as they don't always play well with each other. 

Thank you for this pointer! I had missed it thinking my problem was more general than just lubridate, but this indeed gives nice pointers about S3/S4 classes use inside DT.

Maybe this use of setter functions could be a nice entry in the FAQ? Though again I'm not sure they are very frequent in R, it would make a nice pointer for those wishing to use S3/S4 objects as DT elements

**Comment by avimallu** [comment]
Small correction on your understanding here:

> doesn't work as it seems in that case j returns a copy of the column and not a reference to it (which makes complete sense since i may select random indices and thus the final column cannot be accessed through a single pointer without creating a dedicated array).

You *can* do this in `data.table`. This is called *sub-assign by reference*. See below.

```r
library(data.table)

# Let's build a DT containing a single column with random dates
A=data.table(mydate=as.Date.numeric(
  runif(100,min = 0,max=10000),origin = "2011-01-01"))

A[, test := "test text here"]

A[year(mydate) > 2035, test := "changed text"]

A[test == "changed text"]
```

There are a few nuances you need to be aware of, though, and they are covered in the documentation for `:=`. [This vignette](https://cran.r-project.org/web/packages/data.table/vignettes/datatable-sd-usage.html#column-subsetting-.sdcols) also gives you a good idea of the capabilities using `:=` in `data.table`.

**Comment by qmarcou** [comment]
Sorry if my comment wasn't clear, yes I'm familiar with the *sub-assign by reference* (and that's in fact why I went through the hassle above, because I need to perform the setter operation on subsets of data).

What i meant is:
```r
library(data.table)

# Let's build a DT containing a single column with random dates
A=data.table(mydate=as.Date.numeric(
  runif(100,min = 0,max=10000),origin = "2011-01-01"))

# Subset random indices
some_i_subset = sample(seq_len(A[,.N]),50, replace = TRUE)
A[some_i_subset,.(mydate)][,mydate:=0] # This operation leaves A unchanged
```
while of course using direct *sub-assign by reference* will work

```r
A[some_i_subset, mydate:=0] # Sets about half the dates to 1970-01-01
```
This means `A[some_i_subset,.(mydate)]` returns a copy of the vector containing the `mydate` information, and of course `A[some_i_subset,mydate:=0]` makes no copy. As you suggest my comment about pointers may be too simplistic, I just meant that because `i` can contains arbitrary expressions, including non contiguous and repeated indices (such as `some_i_subset`), keeping a reference and not make a copy of it (while not using *sub-assign by reference*) would be quite complicated (and not desired but that's another topic), since it would mean storing the pointer to each original position of the `mydate` vector contained  in A in the DT created by A[some_i_subset,.(mydate)].

My point above was that because of this behavior (making a copy when `:=` is not used) `A[ ,  mysetterfunction(mydate, value)]` cannot work and `A[ ,  mydate:= mysetterfunction(mydate, value)]` is needed. However the last form implies an unnecessary assignment (and indeed no copy is made, sorry for the confusion in my comment above, I've edited accordingly). What I mean, is that `mysetterfunction(mydate, value)` in the RHS already modifies `mydate` in place and returns a reference (? in the best case, but that's a lubridate discussion) to the initial object, and then the assignment `mydate:=mydate` is performed, which makes an unnecessary assignment by reference. Now I may be wrong about this extra assignment, but then I would be even more amazed by the amazing implementation ;)

This benchmark may be complete nonsense and only reflect memory access but it would say this extra assignment has a cost on large scale (albeit very low):
```r
library(microbenchmark)

A=data.table(mydate=as.Date.numeric(runif(100,min = 0,max=10000),origin = "1970-01-01"))
 microbenchmark::microbenchmark(A[,mydate:=mydate])
Unit: microseconds
                      expr     min      lq     mean   median      uq     max
 A[, `:=`(mydate, mydate)] 228.473 232.047 240.6015 234.1975 237.775 468.603
 neval
   100

# Same on a huge DT
B=data.table(mydate=as.Date.numeric(runif(100000000,min = 0,max=10000),origin = "1970-01-01"))
microbenchmark::microbenchmark(B[,mydate:=mydate])
Unit: milliseconds
                      expr      min       lq     mean   median       uq
 B[, `:=`(mydate, mydate)] 763.9436 783.8511 792.5514 794.9405 798.2983
      max neval
 861.0097   100

# Same assignment with just a scalar
microbenchmark::microbenchmark(B[,mydate:=1])
Unit: milliseconds
                 expr      min       lq     mean   median       uq      max
 B[, `:=`(mydate, 1)] 81.92327 83.35697 83.83702 83.70719 84.10539 88.79019
 neval
   100

# Same self assignment on a DT with 2 columns
 C=data.table(mydate=as.Date.numeric(runif(100000000,min = 0,max=10000),origin = "1970-01-01"),
mydate2=as.Date.numeric(runif(100000000,min = 0,max=10000),origin = "1970-01-01"))
 microbenchmark::microbenchmark(C[,mydate:=mydate])
Unit: milliseconds
                      expr      min       lq     mean   median      uq
 C[, `:=`(mydate, mydate)] 631.4583 670.8538 706.4553 688.7486 716.083
      max neval
 869.2246   100
```
So in a nutshell:
- this self assignment by reference is about 3000 times slower on a column 10^6 times larger
- assignment is about 10 times faster with simple scalar assignment (suggests reading RHS has a cost)
- the time needed for this self assignment by reference is about the same when there are multiple columns
 
Anyway I'm not in this huge data regime and my problem is solved by your previous comments, so this is just some food for thought :)

**Comment by MichaelChirico** [comment]
> In this particular case you may use the functional form of the replacement:
> 
> ```
> A[ , mydate := `mday<-`(mydate, 1)]
> ```

or, without resorting to calling setters directly:

```r
A[, mydate := {
  mday(mydate) = 1
  mydate
}]
```

It looks like the desired syntax here is to interpret `fun(X) := Y` as `` X := `fun<-(X, Y)` ``, assuming such a setter exists, much like the regular R parser converts `fun(X) <- Y` that way, i.e., just replacing the by-reference `LEFT_ASSIGN` with the "normal" left-assign for interpreting the expression.

It's honestly not a bad suggestion, but I'm still not particularly keen to support even more overloading of how to read `:=` expressions. I also think it will start picking up `names<-` setters for the very common existing usage of `names(...) := ...`, so this would either require a lot of special-casing to be back-compatible, or have to be very tightly constrained in ways that sound _ex ante_ hard to maintain.

Given the above workarounds are pretty normal + widely readable, I'll close this as out-of-scope.
